src/datanode/src/heartbeat/handler/upgrade_region.rs

```diff
@@ -0,0 +1,362 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::Duration;
+
+use common_meta::instruction::{InstructionReply, UpgradeRegion, UpgradeRegionReply};
+use common_telemetry::warn;
+use futures_util::future::BoxFuture;
+use store_api::region_request::{RegionCatchupRequest, RegionRequest};
+
+use crate::heartbeat::handler::HandlerContext;
+use crate::heartbeat::task_tracker::WaitResult;
+
+impl HandlerContext {
+    pub(crate) fn handle_upgrade_region_instruction(
+        self,
+        UpgradeRegion {
+            region_id,
+            last_entry_id,
+            wait_for_replay_millis,
+        }: UpgradeRegion,
+    ) -> BoxFuture<'static, InstructionReply> {
+        Box::pin(async move {
+            if let Some(writable) = self.region_server.writable(region_id) {
+                if writable {
+                    return InstructionReply::UpgradeRegion(UpgradeRegionReply {
+                        ready: true,
+                        exists: true,
+                        error: None,
+                    });
+                }
+
+                let region_server_moved = self.region_server.clone();
+
+                // The catchup task is almost zero cost if the inside region is writable.
+                // Therefore, it always registers a new catchup task.
+                let register_result = self
+                    .catchup_tasks
+                    .try_register(
+                        region_id,
+                        Box::pin(async move {
+                            region_server_moved
+                                .handle_request(
+                                    region_id,
+                                    RegionRequest::Catchup(RegionCatchupRequest {
+                                        set_writable: true,
+                                        entry_id: last_entry_id,
+                                    }),
+                                )
+                                .await?;
+
+                            Ok(())
+                        }),
+                    )
+                    .await;
+
+                if register_result.is_busy() {
+                    warn!("Another catchup task is running for the region: {region_id}");
```

Comment:
If busy, shoud we return here?

Reply:
It will wait for the previous running task within a timeout.